### ISSUE TITLE
Sprockets 4 and MIME types

### ISSUE THREAD

**Issue by liseki** [rails/sprockets#273]
I have been trying out Sprockets 4.0.0.beta2 and I have noticed if you have the following app.coffee:
```
#= require constants
```
and ```constants.js``` exists in the load paths, resolve will fail since Sprockets assumes the file directives in ```app.coffee``` will be of MIME type ```text/coffeescript```. I came across this behaviour in my rails app and while running the [teaspoon](https://github.com/modeset/teaspoon/tree/master) project tests.

Is this the intended behaviour? If not, what is the recommended way of going around it or is a patch required?

**Comment by schneems** [comment]
I don't totally follow, can you give me a minimal example app that reproduces the problem?

**Comment by jeremy** [comment]
Requiring constants.js in app.coffee means Sprockets would need to compile JS → Coffee. Technically this is possible, but Sprockets doesn't support it. We could wrap the JS in backticks and call it Coffee. Or support something like js2coffee.

This probably isn't what you expected, though! I wager you're thinking that app.coffee will eventually turn into JS, so requiring a .js file within it will Just Work. But these are separate steps: when you require constants.js in app.coffee, it needs to be transformed to match the .coffee content type. Later on, app.coffee will be requested as app.js and it will be transformed to match the .js content type.

You can address this by using an application.js wrapper than requires your app.coffee and other.js files or by switching your constants.js to constants.coffee and requiring that in app.coffee.

**Comment by schneems** [comment]
What was the error you were getting? Maybe there is a way to better surface and expose the underlying behavior? The application wrapper is a good idea, but maybe not intuitive. I wonder how we could encourage people to do the right thing?

**Comment by liseki** [comment]
Thank you @jeremy and @schneems!

Indeed the wrapper ```application.js``` that Rails uses by default takes of care of that. However I had run into an issue with ```Rails 4.2.5``` and ```Sprockets 4.0.0.beta2``` in the following set up:
```
*** application.js ***
//= require app
```

```
*** app.es6 ***
//= require utils
```

```
*** utils.js ***
...
```

Here Sprockets raised a ```Sprockets::FileNotFound``` error with the message "_couldn't find file 'utils' with type 'application/ecmascript-6'_". In trying to replicate it I noticed this only happened with ```app``` as an ES6 file and not when it was a CoffeeScript file which is what you would expect if this was a general issue with resolving assets.

After a lot of digging I tracked this back to the ```BabelProcessor``` which produced source maps with the wrong source file name. Thanks to @lucasmazza this was [fixed](https://github.com/rails/sprockets/commit/4da21144bb9bed8e73c16ead2b2e0eb225f142be#diff-272bdd84d8054647bee91b8d079de5b2R42) the very next day after the ```4.0.0.beta2``` release.
